src/ipc/Request.h

```diff
@@ -24,19 +25,16 @@ class Request: public RefCountable
     typedef RefCount<Request> Pointer;
 
 public:
-    Request(int aRequestorId, unsigned int aRequestId):
-        requestorId(aRequestorId), requestId(aRequestId) {}
+    Request(int aRequestorId, unsigned int aRequestId, bool initQuid):
+        requestorId(aRequestorId), requestId(aRequestId), qid(initQuid) {}
 
     virtual void pack(TypedMsgHdr& msg) const = 0; ///< prepare for sendmsg()
     virtual Pointer clone() const = 0; ///< returns a copy of this
 
-private:
-    Request(const Request&); // not implemented
-    Request& operator= (const Request&); // not implemented
-
 public:
     int requestorId; ///< kidId of the requestor; used for response destination
     unsigned int requestId; ///< unique for sender; matches request w/ response
+    const QuestionerId qid;
```

Comment:
Unfortunately, I could not make const requestId, because it is assigned in several places (e.g., Ipc::Forwarder::start() and Ipc::Forwarder::swanSong()) - so I had to apply this idea to 'qid' by using const_cast in 'unpack()' methods.

Reply:
> using const_cast in 'unpack()' methods.

I am quickly losing context here, but I am pretty sure that we do not want to lie about `qid` constness just to prevent copying. If we cannot prevent copying naturally because all data members do change, then let's prevent it explicitly by deleting the move assignment operator with a "no assignment of any kind" comment.

Disclaimer: It is not clear to me what exactly we are trying to accomplish by allowing copy construction but prohibiting assignment. However, if we can preserve the current official behavior, then we can move these questions outside the PR scope.

Reply:
I undid the constness trick and added a deleted assignment operator. We cannot delete the move assignment operator (as you suggested) because doing this also [deletes](https://isocpp.org/blog/2018/02/quick-q-most-concise-way-to-disable-copy-and-move-semantics) the copy constructor:

```
../../src/mgr/Response.h:38:5: error: use of deleted function ‘Ipc::Response::Response(const Ipc::Response&)’

```

Reply:
To prohibit assignment, you have to delete both copy assignment operator and move assignment operator. Otherwise, you are only prohibiting some assignments but not others.

One way to prohibit assignment is to explicitly delete just the move assignment because that triggers implicit deletion of the copy assignment operator (among other things). Another way is to explicitly delete both operators. I was suggesting the first way, but the second approach is also OK (and some may reasonably consider it superior when our goal is to preserve the copy construction). The second approach does not need a clarifying comment because it is all explicit.

Yes, deletion of the move assignment operator also deletes the implicitly generated copy constructor (among other things). To enable copy construction, you then _add_ (an explicitly defaulted) copy constructor (and probably an explicitly defaulted move constructor).

It is possible that we want to allow one or both move operations (as an optimization). However, since I do not know what the old code was trying to accomplish here exactly, I would rather preserve the old status quo to the extent possible. AFAICT, that means prohibiting move operations. Please correct me if I am wrong.

The rules for implicit method deletion are very messy, but this is a known weak area in C++. We are not doing anything wrong here AFAICT.

Reply:
The old code, declaring private copy constructor and operator= did not prohibit move operations - in this sense it is almost equivalent to the new code, deleting the only operator= (copy constructors are private in children). So, with current changes, the old behavior is already preserved. We can add a TODO regarding forbidding move operator= (or, at least, clarifying whether it should be done).

Reply:
> The old code, declaring private copy constructor and operator= did not prohibit move operations

Are you sure? I would expect the opposite because the compiler cannot know how to do custom moves, and the existence of a user-declared custom copy constructor implies that custom moves may be required. The well-known table (see page 28 [elsewhere](https://raw.githubusercontent.com/HowardHinnant/presentations/master/accu_2014/accu_2014.pdf)) also seems to suggest that declaring a copy constructor prevents generation of the move members. And the statement under the table on page 25 in the same presentation explicitly says the same.

Reply:
bq. declaring a copy constructor prevents generation of the move members

Agreed.
When I wrote 'did not prohibit move operations' I meant that these members are not implicitly 'deleted' by the compiler (and, as you noted, they are not implicitly declared)  - so a user may is free to declare them. So the point is that the PR does not change the old behavior as far as move operations are concerned.